File: src/main/java/org/noxet/noxetserver/commands/misc/Friend.java
package org.noxet.noxetserver.commands.misc;

import net.md_5.bungee.api.ChatColor;
import org.bukkit.command.Command;
import org.bukkit.command.CommandSender;
import org.bukkit.command.TabExecutor;
import org.bukkit.entity.Player;
import org.noxet.noxetserver.NoxetServer;
import org.noxet.noxetserver.UsernameStorageManager;
import org.noxet.noxetserver.menus.inventory.FriendsMenu;
import org.noxet.noxetserver.messaging.NoxetErrorMessage;
import org.noxet.noxetserver.messaging.NoxetMessage;
import org.noxet.noxetserver.messaging.NoxetNoteMessage;
import org.noxet.noxetserver.messaging.NoxetSuccessMessage;
import org.noxet.noxetserver.playerdata.PlayerDataManager;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;
import java.util.UUID;

public class Friend implements TabExecutor {
    private static final int maxOutgoingRequests = 5, maxIncomingRequests = 20, maxFriends = 30;

    @Override
    @SuppressWarnings("NullableProblems")
    public boolean onCommand(CommandSender commandSender, Command command, String s, String[] strings) {
        if(!(commandSender instanceof Player)) {
            new NoxetErrorMessage(NoxetErrorMessage.ErrorType.COMMON, "Only players can have friends.").send(commandSender);
            return true;
        }

        Player player = (Player) commandSender;

        if(strings.length == 0) {
            new FriendsMenu(player).openInventory(player);
            return true;
        }

        if(strings[0].equalsIgnoreCase("add")) {
            if(strings.length < 2) {
                new NoxetErrorMessage(NoxetErrorMessage.ErrorType.ARGUMENT, "Missing argument: player to befriend.").send(player);
                return true;
            }

            UUID uuidToBefriend = new UsernameStorageManager().getUUIDFromUsernameOrUUID(strings[1]);

            if(uuidToBefriend == null) {
                new NoxetErrorMessage(NoxetErrorMessage.ErrorType.COMMON, "That player is not registered.").send(player);
                return true;
            }

            Player playerToBefriend = NoxetServer.getPlugin().getServer().getPlayer(uuidToBefriend); // Only works if the player is online.
            String befriendName = UsernameStorageManager.getCasedUsernameFromUUID(uuidToBefriend); // This works anyway.

            if(player.getUniqueId().equals(uuidToBefriend)) {
                new NoxetErrorMessage(NoxetErrorMessage.ErrorType.COMMON, "You cannot befriend yourself.").send(player);
                return true;
            }

            if(areFriends(player.getUniqueId(), uuidToBefriend)) {
                new NoxetErrorMessage(NoxetErrorMessage.ErrorType.COMMON, "You are already friends!").send(player);
                return true;
            }

            if(strings.length > 2 && strings[2].equals("!") && player.isOp()) {
                acceptRequest(uuidToBefriend, player.getUniqueId());
                new NoxetSuccessMessage("Forced friendships are the best! You are now friends with " + befriendName + ".").send(player);
                return true;
            }

            if(hasReceivedFriendRequestFrom(uuidToBefriend, player.getUniqueId())) {
                new NoxetErrorMessage(NoxetErrorMessage.ErrorType.COMMON, "You have already sent a friend request to them. Please wait for them to respond.").send(player);
                return true;
            }

            if(getFriendList(player.getUniqueId()).size() >= maxFriends) {
                new NoxetErrorMessage(NoxetErrorMessage.ErrorType.COMMON, "You have reached your friend limit.").send(player);
                return true;
            }

            if(new PlayerDataManager(player).doesContain(PlayerDataManager.Attribute.BLOCKED_PLAYERS, uuidToBefriend.toString())) {
                new NoxetErrorMessage(NoxetErrorMessage.ErrorType.COMMON, "You have blocked this player, and cannot send friend requests to them.").send(player);
                return true;
            }

            if(
                    new PlayerDataManager(uuidToBefriend).doesContain(PlayerDataManager.Attribute.BLOCKED_PLAYERS, player.getUniqueId().toString()) ||
                    (boolean) new PlayerDataManager(uuidToBefriend).get(PlayerDataManager.Attribute.DISALLOW_INCOMING_FRIEND_REQUESTS)
            ) {
                new NoxetErrorMessage(NoxetErrorMessage.ErrorType.COMMON, "You may not send friend requests to this player.").send(player);
                return true;
            }

            if(getFriendList(uuidToBefriend).size() >= maxFriends) {
                new NoxetErrorMessage(NoxetErrorMessage.ErrorType.COMMON, befriendName + " has reached their friend limit!").send(player);
                return true;
            }

            if(hasReceivedFriendRequestFrom(player.getUniqueId(), uuidToBefriend)) {
                acceptRequest(player.getUniqueId(), uuidToBefriend);

                new NoxetSuccessMessage("You are now friends with " + befriendName + "!").send(player);
                if(playerToBefriend != null)
                    new NoxetSuccessMessage(player.getName() + " accepted your friend request.").send(playerToBefriend);

                return true;
            }

            if(getOutgoingFriendRequests(player.getUniqueId()).size() >= maxOutgoingRequests) {
                new NoxetErrorMessage(NoxetErrorMessage.ErrorType.COMMON, "You have reached your limit on max amount of outgoing friend requests.").send(player);
                return true;
            }

            if(getIncomingFriendRequests(uuidToBefriend).size() >= maxIncomingRequests) {
                new NoxetErrorMessage(NoxetErrorMessage.ErrorType.COMMON, befriendName + " has enough friend requests to take care of, let them be!").send(player);
                return true;
            }

            sendFriendRequest(player.getUniqueId(), uuidToBefriend);

            new NoxetMessage("§eYou sent a friend request to " + befriendName + ".").send(player);

            if(playerToBefriend != null)
                new NoxetMessage("§e" + player.getName() + " sent you a friend request!")
                        .addButton("Accept", ChatColor.GREEN, "Become friends!", "friend add " + player.getName())
                        .addButton("Deny", ChatColor.RED, "Don't become friends", "friend deny " + player.getName())
                        .send(playerToBefriend);

            return true;
        }

        if(strings[0].equalsIgnoreCase("remove")) {
            if(strings.length < 2) {
                new NoxetErrorMessage(NoxetErrorMessage.ErrorType.ARGUMENT, "Missing argument: player to unfriend.").send(player);
                return true;
            }

            UUID uuidToUnfriend = new UsernameStorageManager().getUUIDFromUsernameOrUUID(strings[1]);

            if(uuidToUnfriend == null) {
                new NoxetErrorMessage(NoxetErrorMessage.ErrorType.COMMON, "That player is not registered.").send(player);
                return true;
            }

            Player playerToUnfriend = NoxetServer.getPlugin().getServer().getPlayer(uuidToUnfriend); // Only works if the player is online.
            String unfriendName = UsernameStorageManager.getCasedUsernameFromUUID(uuidToUnfriend); // This works anyway.

            if(!areFriends(player.getUniqueId(), uuidToUnfriend)) {
                new NoxetErrorMessage(NoxetErrorMessage.ErrorType.COMMON, "You are not friends!").send(player);
                return true;
            }

            removeFriend(player.getUniqueId(), uuidToUnfriend);

            new NoxetSuccessMessage(unfriendName + " is no longer your friend.").send(player);
            if(playerToUnfriend != null)
                new NoxetMessage("§c" + player.getName() + " removed you as their friend.").send(playerToUnfriend);

            return true;
        }

        if(strings[0].equalsIgnoreCase("deny")) {
            if(strings.length < 2) {
                new NoxetErrorMessage(NoxetErrorMessage.ErrorType.ARGUMENT, "Missing argument: player to deny.").send(player);
                return true;
            }

            UUID uuidToDeny = new UsernameStorageManager().getUUIDFromUsernameOrUUID(strings[1]);

            if(uuidToDeny == null) {
                new NoxetErrorMessage(NoxetErrorMessage.ErrorType.COMMON, "That player is not registered.").send(player);
                return true;
            }

            Player playerToDeny = NoxetServer.getPlugin().getServer().getPlayer(uuidToDeny); // Only works if the player is online.
            String denyName = UsernameStorageManager.getCasedUsernameFromUUID(uuidToDeny); // This works anyway.

            if(!hasReceivedFriendRequestFrom(player.getUniqueId(), uuidToDeny)) {
                new NoxetErrorMessage(NoxetErrorMessage.ErrorType.COMMON, "They have not sent you a friend request!").send(player);
                return true;
            }

            denyRequest(player.getUniqueId(), uuidToDeny);

            new NoxetSuccessMessage("Denied the friend request from " + denyName + ".").send(player);
            if(playerToDeny != null)
                new NoxetMessage("§c" + player.getName() + " denied your friend request.").send(playerToDeny);

            if(strings.length > 2 && strings[2].equals("!"))
                player.performCommand("block " + playerToDeny);

            return true;
        }

        if(strings[0].equalsIgnoreCase("cancel")) {
            if(strings.length < 2) {
                new NoxetErrorMessage(NoxetErrorMessage.ErrorType.ARGUMENT, "Missing argument: player to cancel request to.").send(player);
                return true;
            }

            UUID uuidToCancel = new UsernameStorageManager().getUUIDFromUsernameOrUUID(strings[1]);

            if(uuidToCancel == null) {
                new NoxetErrorMessage(NoxetErrorMessage.ErrorType.COMMON, "That player is not registered.").send(player);
                return true;
            }

            String cancelName = UsernameStorageManager.getCasedUsernameFromUUID(uuidToCancel); // This works anyway.

            if(!hasReceivedFriendRequestFrom(uuidToCancel, player.getUniqueId())) {
                new NoxetErrorMessage(NoxetErrorMessage.ErrorType.COMMON, "You have not sent them a friend request!").send(player);
                return true;
            }

            denyRequest(uuidToCancel, player.getUniqueId());

            new NoxetSuccessMessage("Canceled the friend request to " + cancelName + ".").send(player);

            return true;
        }

        if(strings[0].equalsIgnoreCase("list")) {
            List<String> friendUUIDs = getFriendList(player.getUniqueId());

            new NoxetMessage("§eFriends: " + friendUUIDs.size()).send(player);

            for(String friendUUIDString : friendUUIDs) {
                UUID friendUUID = new UsernameStorageManager().getUUIDFromUsernameOrUUID(friendUUIDString);
                String friendName = UsernameStorageManager.getCasedUsernameFromUUID(friendUUID);

                new NoxetMessage("└§a§lFRIEND §2" + friendName).send(player);
            }

            return true;
        }

        if(strings[0].equalsIgnoreCase("incoming")) {
            List<String> incomingUUIDs = getIncomingFriendRequests(player.getUniqueId());

            new NoxetMessage("§eIncoming friend requests: " + incomingUUIDs.size()).send(player);

            for(String incomingUUIDString : incomingUUIDs) {
                UUID incomingUUID = new UsernameStorageManager().getUUIDFromUsernameOrUUID(incomingUUIDString);
                String incomingName = UsernameStorageManager.getCasedUsernameFromUUID(incomingUUID);

                new NoxetMessage("└§6§lINCOMING §a" + incomingName)
                        .addButton("Accept", ChatColor.GREEN, "Become friends!", "friend add " + incomingUUIDString)
                        .addButton("Deny", ChatColor.RED, "Don't become friends", "friend deny " + incomingUUIDString)
                        .addButton("Deny and block", ChatColor.GRAY, "Deny this request and block " + incomingName, "friend deny " + incomingUUIDString + " !")
                        .send(player);
            }

            return true;
        }

        if(strings[0].equalsIgnoreCase("outgoing")) {
            List<String> outgoingUUIDs = getOutgoingFriendRequests(player.getUniqueId());

            new NoxetMessage("§eOutgoing friend requests: " + outgoingUUIDs.size()).send(player);

            for(String outgoingUUIDString : outgoingUUIDs) {
                UUID outgoingUUID = new UsernameStorageManager().getUUIDFromUsernameOrUUID(outgoingUUIDString);
                String outgoingName = UsernameStorageManager.getCasedUsernameFromUUID(outgoingUUID);

                new NoxetMessage("└§8§lOUTGOING §7" + outgoingName)
                        .addButton("Cancel", ChatColor.RED, "Regret wanting to become friends?", "friend cancel " + outgoingUUIDString)
                        .send(player);
            }

            return true;
        }

        if(strings[0].equalsIgnoreCase("toggle-allow-incoming")) {
            PlayerDataManager playerDataManager = new PlayerDataManager(player);
            boolean setTo = !(boolean) playerDataManager.get(PlayerDataManager.Attribute.DISALLOW_INCOMING_FRIEND_REQUESTS);
            playerDataManager.set(PlayerDataManager.Attribute.DISALLOW_INCOMING_FRIEND_REQUESTS, setTo).save();

            new NoxetSuccessMessage((setTo ? "Disabled" : "Enabled") + " incoming friend requests. Other players can " + (setTo ? "no longer" : "now") + " send friend requests to you.").send(player);

            if(setTo && playerDataManager.getListSize(PlayerDataManager.Attribute.INCOMING_FRIEND_REQUESTS) != 0)
                new NoxetNoteMessage("You have incoming friend requests already pending. If you want, you can cancel those manually.")
                        .addButton("Manage", ChatColor.YELLOW, "Manage incoming friend requests", "friend incoming")
                        .send(player);

            return true;
        }

        if(strings[0].equalsIgnoreCase("toggle-friend-tp")) {
            PlayerDataManager playerDataManager = new PlayerDataManager(player);
            boolean setTo = !(boolean) playerDataManager.get(PlayerDataManager.Attribute.FRIEND_TELEPORTATION);
            playerDataManager.set(PlayerDataManager.Attribute.FRIEND_TELEPORTATION, setTo).save();

            new NoxetSuccessMessage((setTo ? "Enabled" : "Disabled") + " friend teleportation. Your friends can " + (setTo ? "now" : "no longer") + " /tpa to you without you accepting.").send(player);

            return true;
        }

        new NoxetErrorMessage(NoxetErrorMessage.ErrorType.ARGUMENT, "Sorry, '" + strings[0] + "' is not a valid friend command.").send(player);

        return true;
    }

    @Override
    @SuppressWarnings("NullableProblems")
    public List<String> onTabComplete(CommandSender commandSender, Command command, String s, String[] strings) {
        if(!(commandSender instanceof Player))
            return null;

        List<String> completions = new ArrayList<>();

        Player player = (Player) commandSender;

        if(strings.length == 1) {
            completions.addAll(Arrays.asList("add", "remove", "deny", "list", "incoming", "outgoing", "toggle-allow-incoming", "toggle-friend-tp"));
        } else if(strings.length == 2) {
            switch(strings[0].toLowerCase()) {
                case "add":
                    Player playerToRecommend = NoxetServer.getPlugin().getServer().getPlayer(strings[1]);
                    if(playerToRecommend != null)
                        completions.add(playerToRecommend.getName());
                case "deny":
                    for(String incomingUUIDString : getIncomingFriendRequests(player.getUniqueId())) {
                        UUID incomingUUID = new UsernameStorageManager().getUUIDFromUsernameOrUUID(incomingUUIDString);
                        completions.add(incomingUUID != null ? UsernameStorageManager.getCasedUsernameFromUUID(incomingUUID) : incomingUUIDString);
                    }

                    break;
                case "remove":
                    for(String friendUUIDString : getFriendList(player.getUniqueId())) {
                        UUID friendUUID = new UsernameStorageManager().getUUIDFromUsernameOrUUID(friendUUIDString);
                        completions.add(friendUUID != null ? UsernameStorageManager.getCasedUsernameFromUUID(friendUUID) : friendUUIDString);
                    }

                    break;
                case "cancel":
                    for(String outgoingUUIDString : getOutgoingFriendRequests(player.getUniqueId())) {
                        UUID outgoingUUID = new UsernameStorageManager().getUUIDFromUsernameOrUUID(outgoingUUIDString);
                        completions.add(outgoingUUID != null ? UsernameStorageManager.getCasedUsernameFromUUID(outgoingUUID) : outgoingUUIDString);
                    }

                    break;
            }
        }

        return completions;
    }

    public static void sendFriendRequest(UUID from, UUID to) {
        new PlayerDataManager(from).addToStringList(PlayerDataManager.Attribute.OUTGOING_FRIEND_REQUESTS, to.toString()).save();
        new PlayerDataManager(to).addToStringList(PlayerDataManager.Attribute.INCOMING_FRIEND_REQUESTS, from.toString()).save();
    }

    public static boolean areFriends(UUID player1, UUID player2) {
        return getFriendList(player1).contains(player2.toString()) &&
                getFriendList(player2).contains(player1.toString());
    }

    @SuppressWarnings("unchecked")
    public static List<String> getFriendList(UUID uuid) {
        return (List<String>) new PlayerDataManager(uuid).get(PlayerDataManager.Attribute.FRIEND_LIST);
    }

    @SuppressWarnings("unchecked")
    public static List<String> getOutgoingFriendRequests(UUID uuid) {
        return (List<String>) new PlayerDataManager(uuid).get(PlayerDataManager.Attribute.OUTGOING_FRIEND_REQUESTS);
    }

    @SuppressWarnings("unchecked")
    public static List<String> getIncomingFriendRequests(UUID uuid) {
        return (List<String>) new PlayerDataManager(uuid).get(PlayerDataManager.Attribute.INCOMING_FRIEND_REQUESTS);
    }

    public static boolean hasReceivedFriendRequestFrom(UUID to, UUID from) {
        return getIncomingFriendRequests(to).contains(from.toString());
    }

    public static void acceptRequest(UUID to, UUID from) {
        new PlayerDataManager(to)
                .removeFromStringList(PlayerDataManager.Attribute.INCOMING_FRIEND_REQUESTS, from.toString())
                .addToStringList(PlayerDataManager.Attribute.FRIEND_LIST, from.toString())
                .save();
        new PlayerDataManager(from)
                .removeFromStringList(PlayerDataManager.Attribute.OUTGOING_FRIEND_REQUESTS, to.toString())
                .addToStringList(PlayerDataManager.Attribute.FRIEND_LIST, to.toString())
                .save();
    }

    public static void denyRequest(UUID to, UUID from) {
        new PlayerDataManager(to)
                .removeFromStringList(PlayerDataManager.Attribute.INCOMING_FRIEND_REQUESTS, from.toString())
                .save();
        new PlayerDataManager(from)
                .removeFromStringList(PlayerDataManager.Attribute.OUTGOING_FRIEND_REQUESTS, to.toString())
                .save();
    }

    public static void removeFriend(UUID player1, UUID player2) {
        new PlayerDataManager(player1)
                .removeFromStringList(PlayerDataManager.Attribute.FRIEND_LIST, player2.toString())
                .save();
        new PlayerDataManager(player2)
                .removeFromStringList(PlayerDataManager.Attribute.FRIEND_LIST, player1.toString())
                .save();
    }
}

File: src/main/java/org/noxet/noxetserver/commands/social/Friend.scala
package org.noxet.noxetserver.commands.social

import org.bukkit.command.{Command, CommandSender, TabExecutor}
import org.bukkit.entity.Player
import org.noxet.noxetserver.NoxetServer
import org.noxet.noxetserver.commands.RegisteredCommand
import org.noxet.noxetserver.menus.inventory.FriendsMenu
import org.noxet.noxetserver.messaging.{ErrorMessage, Message, NoteMessage, SuccessMessage}
import org.noxet.noxetserver.playerdata.PlayerDataManager
import org.noxet.noxetserver.playerdata.PlayerDataManager.Attribute.*
import org.noxet.noxetserver.util.{PlayerIntel, UsernameStorageManager}

import java.util.UUID
import scala.collection.immutable.{AbstractSeq, LinearSeq}

object Friend extends PlayerTabExecutor, RegisteredCommand("friend", this):
  val MaxOutgoingRequests = 5
  val MaxIncomingRequests = 20
  val MaxFriends = 30

  override def onPlayerCommand(player: Player, command: Command, s: String, strings: Array[String]): Boolean =
    strings match
      case Array() =>
        FriendsMenu(player).openInventory(player)
        true
      case Array("add") =>
        ErrorMessage(ErrorMessage.ErrorType.Argument, "Missing argument: player to befriend.").send(player)
        true
      case Array("add", playerIdentifier, args*) =>
        val befriend = PlayerFragments(playerIdentifier) match
          case Some(x) => x
          case None =>
            ErrorMessage(ErrorMessage.ErrorType.Common, "That player is not registered.").send(player)
            return true

        if player.getUniqueId == befriend.uuid then
          ErrorMessage(ErrorMessage.ErrorType.Common, "You cannot befriend yourself.").send(player)
          return true

        if areFriends(player, befriend) then
          ErrorMessage(ErrorMessage.ErrorType.Common, "You are already friends!").send(player)
          return true

        args.headOption match
          case Some("!") if player.isOp =>
            acceptRequest(befriend, player)
            SuccessMessage(s"Forced friendships are the best! You are now friends with $befriend.").send(player)
            return true
          case _ => ()

        if hasReceivedFriendRequestFrom(befriend, player) then
          ErrorMessage(ErrorMessage.ErrorType.Common, "You have already sent a friend request to them. Please wait for them to respond.").send(player)
          return true

        if getFriendList(player).size >= maxFriends then
          ErrorMessage(ErrorMessage.ErrorType.Common, "You have reached your friend limit.").send(player)
          return true

        if PlayerDataManager(player).doesContain(BlockedPlayers, befriend.uuid.toString) then
          ErrorMessage(ErrorMessage.ErrorType.Common, "You have blocked this player, and cannot send friend requests to them.").send(player)
          return true

        if PlayerDataManager(befriend).doesContain(BlockedPlayers, player.getUniqueId.toString) ||
          PlayerDataManager(befriend).get(DisallowIncomingFriendRequests).toBoolean
        then
          ErrorMessage(ErrorMessage.ErrorType.Common, "You may not send friend requests to this player.").send(player)
          return true

        if getFriendList(befriend).size >= maxFriends then
          ErrorMessage(ErrorMessage.ErrorType.Common, s"$befriend has reached their friend limit!").send(player)
          return true

        if hasReceivedFriendRequestFrom(player, befriend) then
          acceptRequest(player, befriend)

          SuccessMessage(s"You are now friends with $befriend!").send(player)
          befriend.playerOption.foreach(SuccessMessage(s"${player.getName} accepted your friend request.").send(_))

          return true

        if getOutgoingFriendRequests(player).size >= MaxOutgoingRequests then
          ErrorMessage(ErrorMessage.ErrorType.Common, "You have reached your limit on max amount of outgoing friend requests.").send(player)
          return true

        if getIncomingFriendRequests(befriend).size >= MaxIncomingRequests then
          ErrorMessage(ErrorMessage.ErrorType.Common, s"$befriend has enough friend requests to take care of, let them be!").send(player)
          return true

        sendFriendRequest(player, befriend)

        Message(s"§eYou sent a friend request to $befriend.").send(player)

        befriend.playerOption.foreach(
          Message(s"§e${player.getName} sent you a friend request!")
            .addButton("Accept", ChatColor.GREEN, "Become friends!", s"friend add ${player.getName}")
            .addButton("Deny", ChatColor.RED, "Don't become friends", s"friend deny ${player.getName}")
            .send)

        true
      case Array("remove") =>
        ErrorMessage(ErrorMessage.ErrorType.Argument, "Missing argument: player to unfriend.").send(player)
        true
      case Array("remove", playerIdentifier, args*) =>
        val unfriend = PlayerFragments(playerIdentifier) match
          case Some(x) => x
          case None =>
            ErrorMessage(ErrorMessage.ErrorType.Common, "That player is not registered.").send(player)
            return true

        if !areFriends(player, unfriend) then
          ErrorMessage(ErrorMessage.ErrorType.Common, "You are not friends!").send(player)
          return true

        removeFriend(player, unfriend)

        SuccessMessage(s"$unfriend is no longer your friend.").send(player)
        unfriend.playerOption.foreach(Message(s"§c${player.getName} removed you as their friend.").send(_))

        true
      case Array("deny") =>
        ErrorMessage(ErrorMessage.ErrorType.Argument, "Missing argument: player to deny.").send(player)
        true
      case Array("deny", playerIdentifier, args*) =>
        val denyPlayer = PlayerFragments(playerIdentifier) match
          case Some(x) => x
          case None =>
            ErrorMessage(ErrorMessage.ErrorType.Common, "That player is not registered.").send(player)
            return true

        if !hasReceivedFriendRequestFrom(player, denyPlayer) then
          ErrorMessage(ErrorMessage.ErrorType.Common, "They have not sent you a friend request!").send(player)
          return true

        denyRequest(player, denyPlayer)

        SuccessMessage(s"Denied the friend request from $denyPlayer.").send(player)
        denyPlayer.playerOption.foreach(Message(s"§c${player.getName} denied your friend request.").send)

        args.headOption match
          case Some("!") =>
            player.performCommand(s"block ${denyPlayer.username}")
          case _ => ()

        true
      case Array("cancel") =>
        ErrorMessage(ErrorMessage.ErrorType.Argument, "Missing argument: player to cancel request to.").send(player)
        true
      case Array("cancel", playerIdentifier, args*) =>
        val cancelTo = PlayerFragments(playerIdentifier) match
          case Some(x) => x
          case None =>
            ErrorMessage(ErrorMessage.ErrorType.Common, "That player is not registered.").send(player)
            return true

        if !hasReceivedFriendRequestFrom(cancelTo, player) then
          ErrorMessage(ErrorMessage.ErrorType.Common, "You have not sent them a friend request!").send(player)
          return true

        denyRequest(cancelTo, player)

        SuccessMessage(s"Canceled the friend request to $cancelTo.").send(player)

        true
      case Array("list", _*) =>
        val friendUuids = getFriendList(player.getUniqueId)
        Message(s"§eFriends: ${friendUuids.size}").send(player)
        friendUuids.map(PlayerIntel).foreach(friend => Message(s"└§a§lFRIEND §2$friend").send(player))
        true
      case Array("incoming", _*) =>
        val incomingUuids = getIncomingFriendRequests(player.getUniqueId)
        Message(s"§eIncoming friend requests: ${incomingUuids.size}").send(player)
        incomingUuids.map(PlayerIntel).foreach(incomingFriend =>
          Message(s"└§6§lINCOMING §a${incomingFriend.username}")
            .addButton("Accept", ChatColor.GREEN, "Become friends!", s"friend add $incomingFriend")
            .addButton("Deny", ChatColor.RED, "Don't become friends", s"friend deny $incomingFriend")
            .addButton("Deny and block", ChatColor.GRAY, s"Deny this request and block $incomingFriend", s"friend deny $incomingFriend !")
            .send(player))
        true
      case Array("outgoing", _*) =>
        val outgoingUuids = getOutgoingFriendRequests(player)
        Message(s"§eOutgoing friend requests: ${outgoingUuids.size}").send(player)
        outgoingUuids.map(PlayerIntel).foreach(outgoingFriend =>
          Message(s"└§8§lOUTGOING §7$outgoingFriend")
            .addButton("Cancel", ChatColor.RED, "Regret wanting to become friends?", s"friend cancel $outgoingFriend")
            .send(player))
        true
      case Array("toggle-allow-incoming", _*) =>
        val playerDataManager = PlayerDataManager(player)
        val setTo = !playerDataManager.get(DisallowIncomingFriendRequests).toBoolean
        playerDataManager.set(DisallowIncomingFriendRequests, setTo).save()
        SuccessMessage(
          (if setTo then "Disabled" else "Enabled") +
            " incoming friend requests. Other players can " +
            (if setTo then "no longer" else "now") +
            " send friend requests to you.").send(player)
        if setTo && playerDataManager.getListSize(IncomingFriendRequests) != 0 then
          NoteMessage("You have incoming friend requests already pending. If you want, you can cancel those manually.")
            .addButton("Manage", ChatColor.YELLOW, "Manage incoming friend requests", "friend incoming")
            .send(player)
        true
      case Array("toggle-friend-tp", _*) =>
        val playerDataManager = PlayerDataManager(player)
        val setTo = !playerDataManager.get(FriendTeleportation).toBoolean
        playerDataManager.set(FriendTeleportation, setTo).save()
        SuccessMessage(
          s"${if setTo then "Enabled" else "Disabled"} friend teleportation." +
            s"Your friends can ${if setTo then "now" else "no longer"} /tpa to you without you accepting.").send(player)
        true
      case Array(action, _*) =>
        ErrorMessage(ErrorMessage.ErrorType.Argument, s"Sorry, '$action' is not a valid friend command.").send(player)
        true

  override def onPlayerTabComplete(player: Player, command: Command, s: String, strings: Array[String]): List[String] =
    strings match
      case Array(_) =>
        List("add", "remove", "deny", "list", "incoming", "outgoing", "toggle-allow-incoming", "toggle-friend-tp")
      case Array(action, arg) =>
        action match
          case "add" =>
            Option(NoxetServer.getPlugin.getServer.getPlayer(arg))
              .map(_.getName)
              .toList
          case "deny" =>
            getIncomingFriendRequests(player)
              .map(PlayerIntel)
              .map(_.username)
          case "remove" =>
            getFriendList(player)
              .map(PlayerIntel)
              .map(_.username)
          case "cancel" =>
            getOutgoingFriendRequests(player)
              .map(PlayerIntel)
              .map(_.username)
      case _ => List()

  def sendFriendRequest(from: UUID, to: UUID): Unit =
    PlayerDataManager(from).addToStringList(OutgoingFriendRequests, to.toString).save()
    PlayerDataManager(to).addToStringList(IncomingFriendRequests, from.toString).save()

  def areFriends(player1: UUID, player2: UUID): Boolean =
    getFriendList(player1).contains(player2.toString) &&
      getFriendList(player2).contains(player1.toString)

  def getFriendList(uuid: UUID): List[String] =
    PlayerDataManager(uuid).get(FriendList).asInstanceOf[List[String]]

  def getOutgoingFriendRequests(uuid: UUID): List[String] =
    PlayerDataManager(uuid).get(OutgoingFriendRequests).asInstanceOf[List[String]]

  def getIncomingFriendRequests(uuid: UUID): List[String] =
    PlayerDataManager(uuid).get(IncomingFriendRequests).asInstanceOf[List[String]]

  def hasReceivedFriendRequestFrom(to: UUID, from: UUID): Boolean =
    getIncomingFriendRequests(to).contains(from.toString)

  def acceptRequest(to: UUID, from: UUID): Unit =
    PlayerDataManager(to)
      .removeFromStringList(IncomingFriendRequests, from.toString)
      .addToStringList(FriendList, from.toString)
      .save()
    PlayerDataManager(from)
      .removeFromStringList(OutgoingFriendRequests, to.toString)
      .addToStringList(FriendList, to.toString)
      .save()

  def denyRequest(to: UUID, from: UUID): Unit =
    PlayerDataManager(to)
      .removeFromStringList(IncomingFriendRequests, from.toString)
      .save()
    PlayerDataManager(from)
      .removeFromStringList(OutgoingFriendRequests, to.toString)
      .save()

  def removeFriend(player1: UUID, player2: UUID): Unit =
    PlayerDataManager(player1)
      .removeFromStringList(FriendList, player2.toString)
      .save()
    PlayerDataManager(player2)
      .removeFromStringList(FriendList, player1.toString)
      .save()